pkg/driver/driver_azure.go

```diff
@@ -1035,20 +1047,64 @@ func (clients *azureDriverClients) checkOrphanDisks(ctx context.Context, resourc
 }
 
 func (clients *azureDriverClients) deleteNIC(ctx context.Context, resourceGroupName string, nicName string) error {
+	var (
+		nicDeletionTimeout     = 10 * time.Minute
+		nicDeletionCtx, cancel = context.WithTimeout(ctx, nicDeletionTimeout)
+	)
+	defer cancel()
+
 	klog.V(2).Infof("NIC delete started for %q", nicName)
 	defer klog.V(2).Infof("NIC deleted for %q", nicName)
 
-	future, err := clients.nic.Delete(ctx, resourceGroupName, nicName)
+	future, err := clients.nic.Delete(nicDeletionCtx, resourceGroupName, nicName)
 	if err != nil {
 		return onARMAPIErrorFail(prometheusServiceNIC, err, "nic.Delete")
 	}
-	if err := future.WaitForCompletionRef(ctx, clients.nic.Client); err != nil {
+	if err := future.WaitForCompletionRef(nicDeletionCtx, clients.nic.Client); err != nil {
+		return onARMAPIErrorFail(prometheusServiceNIC, err, "nic.Delete")
+	}
+
+	if err := clients.waitUntilNICIsDeleted(nicDeletionCtx, resourceGroupName, nicName); err != nil {
 		return onARMAPIErrorFail(prometheusServiceNIC, err, "nic.Delete")
 	}
+
 	onARMAPISuccess(prometheusServiceNIC, "NIC deletion was successful for %s", nicName)
 	return nil
 }
 
+func (clients *azureDriverClients) waitUntilNICIsDeleted(ctx context.Context, resourceGroupName string, nicName string) error {
+	var (
+		sleepInterval    = 10 * time.Second
+		maxSleepInterval = 3 * time.Minute
+		currentSleepTime = 0 * time.Second
```

Comment:
Constants? Or even better, why not use client-go/backoff? https://github.com/kubernetes/client-go/blob/master/util/flowcontrol/backoff.go#L32-L38

Reply:
I was actually thinking of using a backoff client, didn't know which one to use. Yes, this makes sense. Let me check it out.

Reply:
I looked through using the client-go/backoff mechanism. However, it looks like it's usecase is a little different compared to ours. It seems like it's keeping a bucket for objects with different names that are being backed off. Refer below
- https://github.com/kubernetes/client-go/blob/master/util/flowcontrol/backoff.go#L37
- https://github.com/kubernetes/kubernetes/blob/019080fc4d32bc7acc994605afbc2fac416188d7/pkg/controller/daemon/daemon_controller.go#L820-L838

So looked up more appropriate libraries for our use-case. I found [this backoff library](https://github.com/cenkalti/backoff) that seems to be used by a lot of projects. And usage also seems simple and works well in our case. The Licence however is MIT, I hope that should be fine as I see other vendors using similar licenses - https://github.com/cenkalti/backoff/blob/v4/LICENSE.